crates/math/src/multiply_fraction.rs

```diff
@@ -13,3 +15,64 @@ where
 
     fn checked_div_dec_ceil(self, rhs: F) -> MathResult<Self>;
 }
+
+impl<U, AsU, F> MultiplyFraction<F, AsU> for Uint<U>
+where
+    Uint<U>: NumberConst + Number + IsZero + MultiplyRatio + From<Uint<AsU>> + ToString,
+    F: Number + Fraction<AsU> + Sign + ToString + IsZero,
+{
+    fn checked_mul_dec_floor(self, rhs: F) -> MathResult<Self> {
+        // If either left or right hand side is zero, then simply return zero.
+        if self.is_zero() || rhs.is_zero() {
+            return Ok(Self::ZERO);
+        }
+        // The left hand side is `Uint`, a non-negative type, so multiplication
+        // with any non-zero negative number goes out of bound.
+        if rhs.is_negative() {
+            return Err(MathError::negative_mul(self, rhs));
+        }
+        self.checked_multiply_ratio_floor(rhs.numerator(), F::denominator())
+    }
+
+    fn checked_mul_dec_ceil(self, rhs: F) -> MathResult<Self> {
+        if self.is_zero() || rhs.is_zero() {
+            return Ok(Self::ZERO);
+        }
+        if rhs.is_negative() {
+            return Err(MathError::negative_mul(self, rhs));
+        }
+        self.checked_multiply_ratio_ceil(rhs.numerator(), F::denominator())
+    }
+
+    fn checked_div_dec_floor(self, rhs: F) -> MathResult<Self> {
+        // If right hand side is zero, throw error, because you can't divide any
+        // number by zero.
+        if rhs.is_zero() {
+            return Err(MathError::division_by_zero(self));
+        }
+        // If right hand side is negative, throw error, because you can't divide
+        // and unsigned number with a negative number.
+        if rhs.is_negative() {
+            return Err(MathError::negative_div(self, rhs));
+        }
+        // If left hand side is zero, and we know right hand size is positive,
+        // then simply return zero.
+        if self.is_zero() {
+            return Ok(Self::ZERO);
+        }
+        self.checked_multiply_ratio_floor(F::denominator(), rhs.numerator())
+    }
+
+    fn checked_div_dec_ceil(self, rhs: F) -> MathResult<Self> {
+        if rhs.is_zero() {
+            return Err(MathError::division_by_zero(self));
+        }
+        if rhs.is_negative() {
+            return Err(MathError::negative_div(self, rhs));
+        }
+        if self.is_zero() {
+            return Ok(Self::ZERO);
+        }
+        self.checked_multiply_ratio_ceil(F::denominator(), rhs.numerator())
+    }
+}
```

Comment:
Just noticed, are these negative check a thing?

Reply:
`Uint` was intended for only positive numbers, so we have it. Now I think it can be removed